compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1918,6 +1918,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         receiver: Option<&'tcx hir::Expr<'tcx>>,
         args: &'tcx [hir::Expr<'tcx>],
     ) -> bool {
+        // Do not call `fn_sig` on non-functions.
+        if !matches!(
+            self.tcx.def_kind(def_id),
+            DefKind::Fn | DefKind::AssocFn | DefKind::Variant | DefKind::Ctor(..)
```

Comment:
Hm, what about closures? Side-note, there's [`DefKind::is_fn_like`](https://doc.rust-lang.org/beta/nightly-rustc/src/rustc_hir/def.rs.html#229-234), but it's slightly different that this case... Maybe worth checking use-cases and syncing them up.

Reply:
Wait, closures make no sense -- they have no where clauses, lol.

Reply:
Calling `fn_sig` on closures and generators ICEs too.